File: server/datasets/controlled_benchmarks/python/multi_student_cohorts/cohort_a3/organic_4.py
# Linked List Reversal - organic submission 4
# "Convert to array, reverse the array, rebuild the list" approach - genuinely
# different technique from pointer-swapping.

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


def build_linked_list(items):
    dummy = Node(None)
    tail = dummy
    for item in items:
        tail.next = Node(item)
        tail = tail.next
    return dummy.next


def linked_list_to_array(head):
    arr = []
    node = head
    while node is not None:
        arr.append(node.data)
        node = node.next
    return arr


def reverse_via_array(head):
    arr = linked_list_to_array(head)
    arr.reverse()
    return build_linked_list(arr)


def show(head):
    node = head
    parts = []
    while node is not None:
        parts.append(str(node.data))
        node = node.next
    print(", ".join(parts))


if __name__ == "__main__":
    lst = build_linked_list([2, 4, 6, 8, 10])
    show(lst)
    lst = reverse_via_array(lst)
    show(lst)

File: server/datasets/controlled_benchmarks/java/multiple_files/multiple_a3/organic_4.java
// Linked List Reversal - organic submission 4
// Convert to array, reverse array, rebuild list.

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class Node {
    int data;
    Node next;

    public Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class Solution {
    public static Node buildLinkedList(int[] items) {
        Node dummy = new Node(0);
        Node cur = dummy;
        for (int item : items) {
            cur.next = new Node(item);
            cur = cur.next;
        }
        return dummy.next;
    }

    public static Node reverseViaArray(Node head) {
        List<Integer> values = new ArrayList<>();
        Node cur = head;
        while (cur != null) {
            values.add(cur.data);
            cur = cur.next;
        }
        Collections.reverse(values);
        int[] arr = new int[values.size()];
        for (int i = 0; i < values.size(); i++) arr[i] = values.get(i);
        return buildLinkedList(arr);
    }

    public static void printNodes(Node head) {
        Node cur = head;
        StringBuilder sb = new StringBuilder();
        while (cur != null) {
            sb.append(cur.data);
            if (cur.next != null) sb.append(" -> ");
            cur = cur.next;
        }
        System.out.println(sb.toString());
    }

    public static void main(String[] args) {
        Node head = buildLinkedList(new int[]{5, 4, 3, 2, 1});
        printNodes(head);
        head = reverseViaArray(head);
        printNodes(head);
    }
}
